lib/puppet_x/templates/invoke_ps_command.erb

```
@@ -1,15 +1,255 @@
+    {
+      throw new NotImplementedException();
+    }
+  }
+
+  public class PowerShellHostUserInterface : PSHostUserInterface
+  {
+    private PowerShellRawUserInterface _rawui;
+    private StringBuilder _sb;
+
+    public PowerShellHostUserInterface()
+    {
+      _sb = new StringBuilder();
+    }
+
+    public override PSHostRawUserInterface RawUI
+    {
+      get
+      {
+        if ( _rawui == null){
+          _rawui = new PowerShellRawUserInterface();
+        }
+        return _rawui;
+      }
+    }
+
+    public override void Write(ConsoleColor foregroundColor, ConsoleColor backgroundColor, string value)
+    {
+      _sb.Append(value);
+    }
+
+    public override void Write(string value)
+    {
+      _sb.Append(value);
+    }
+
+    public override void WriteDebugLine(string message)
+    {
+      _sb.AppendLine("DEBUG: " + message);
+    }
+
+    public override void WriteErrorLine(string value)
+    {
+      _sb.AppendLine(value);
+    }
+
+    public override void WriteLine(string value)
+    {
+      _sb.AppendLine(value);
+    }
+
+    public override void WriteVerboseLine(string message)
+    {
+      _sb.AppendLine("VERBOSE: " + message);
+    }
+
+    public override void WriteWarningLine(string message)
+    {
+      _sb.AppendLine("WARNING: " + message);
+    }
+
+    private bool hasLoggedStartProgress = false;
+    private bool hasLoggedFinalProgress = false;
+    public override void WriteProgress(long sourceId, ProgressRecord record)
+    {
+      if(record.PercentComplete == -1) return;
+      if(hasLoggedFinalProgress) return;
+      if(!hasLoggedStartProgress)
+      {
+        hasLoggedStartProgress = true;
+        _sb.AppendLine(record.Activity.ToString());
+      }
+
+      // http://stackoverflow.com/a/888569/18475
+      var msg = string.Format("\rProgress: {0}% - {1}", record.PercentComplete.ToString(), record.StatusDescription.PadRight(50, ' ') );
+      _sb.Append(msg);
```

The v1 Powershell Module behaviour was to NOT write the progress to the stdout I'm happy with silently ignoring Write-Progress